deliverylocation/models.py

```diff
@@ -0,0 +1,146 @@
+    user_type = models.CharField(max_length=16,
+                                 validators=[MinLengthValidator(6)],
+                                 null=False,
+                                 choices=[('client', 'client'), ('supplier', 'supplier')],
+                                 help_text="User Type - supplier/client")
+
+    class Meta:
+        abstract = True
+
+
+class Supplier(User):
+    """Supplier class inherit from User abstract class.
+
+    Supplier class lets the app user to do variety of actions
+    which evolves the Product and Delivery class.
+
+    Attributes:
+        business_name: a string which holds the Supplier business name.
+    """
+    business_name = models.CharField(max_length=32,
+                                     validators=[MinLengthValidator(6)],
+                                     null=False,
+                                     help_text="Business Name")
+
+    @staticmethod
+    def filter_by_field(field, field_value):
+        """Fetches list of Suppliers filtered with given field and given value.
+
+        Args:
+        field:
+            Supplier attribute.
+        field_value:
+            value which will filter the suppliers.
+
+        Returns:
+            returns a queryset containing the objects that match the specified filter conditions.
+
+        Raises:
+        FieldDoesNotExist error: if Supplier doesn't have the specified attribute.
+        """
+        if hasattr(Supplier, str(field)):
+            return Supplier.objects.filter(**{field: field_value})
+        raise FieldDoesNotExist
+
+    @staticmethod
+    def get_all_suppliers():
+        """Fetches list of all Suppliers.
+
+        Args:
+            None.
+
+        Returns:
+            returns a queryset containing all the Suppliers.
+        """
+        return Supplier.objects.all()
+
+    @staticmethod
+    def get_all_fields():
+        """Fetches list of all Supplier fields.
+
+        Args:
+            None.
+
+        Returns:
+            returns a list of all Supplier fields.
+        """
+        return ['user_name', 'first_name', 'last_name',
+                'password', 'user_type', 'business_name']
+
+
+class DeliveryLocation(models.Model):
+    user_name = models.ForeignKey(Supplier, on_delete=models.CASCADE)
+    location = models.CharField(max_length=32, validators=[MinLengthValidator(1)])
+    date = models.DateField()
+
+    def add_delivery_location(self):
+        self.save()
+        try:
+            DeliveryLocation.full_clean(self)
+        except ValidationError as e:
+            self.delete()
+            raise e
+        return self
+
+    def remove_delivery_location(self):
+        DeliveryLocation.objects.get(id=self.id).delete()
+
+    def update_delivery_location_date(self, newdate):
+        self.date = newdate
+        self.save()
+
+    def update_delivery_location_location(self, newlocation):
+        if len(newlocation) >= 1:
```

Comment:
Do we want to ignore illegal input here and just silently skip? Can we instead remove the `if` and call `full_clean` or something similar?

Reply:
newlocation is a string,if illegal input must be something like null which doesnt have len >=1

Reply:
`len(None)` will raise an exception. You can use instead:
```
if newlocation:
```
but still, I don't understand the logic. having an update function that fails silently looks risky -- or at least bad user experience (imagine you update a value and it stays the way it was without letting you know why).

Reply:
Updated(Also the logic that the new location is bigger than zero going to be in frontend pages too)
i have added if len = 0 or newlocation is None validationerror will be raised with specific message,other than that it will succeed